### PR TITLE
[WIP] Make conftests generated and/or built files have unique names - Issue#3469

### PR DESCRIPTION
**Pull request by acmorrow**
Address issue #3469 

> If a command line option can cause a different sequence of configure checks to run, or to run in a different order, SCons can become confused about which tests are cached, or about which actions are associated with a given numbered configure check. This can lead to either re-running configure checks that should have still been cached, only because the numbering changed, or, in worse cases, to a misconfigured build. See https://jira.mongodb.org/browse/SERVER-23208 for an example.

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [ ] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by acmorrow** [comment]
This is some WIP work I had for #3469. I'll leave some inline comments explaining in more detail.

File: src/engine/SCons/SConf.py [hunk 1]
```diff
@@ -592,8 +593,30 @@ def TryBuild(self, builder, text = None, extension = ""):
             raise SCons.Errors.UserError('Missing SPAWN construction variable.')
 
         nodesToBeBuilt = []
+        sourcetext = self.env.Value(text)
+        f = "conftest"
+
+        if text is not None:
+            textSig = SCons.Util.MD5signature(sourcetext)
+            textSigCounter = str(_ac_build_counter[textSig])
+            _ac_build_counter[textSig] += 1
```
Comment (acmorrow) [inline src/engine/SCons/SConf.py:602]: I'd been hoping to entirely eliminate the counter. But without it, you end up with many configure tests which have the same text body (notably, the empty body). If those configure tests came from different environments, you get the dreaded "scons: warning: Two different environments were specified for target" issue. So, we re-purpose the counter to differentiate identical files. That is sort of unfortunate, but it still works, because if the numbering ever changes in a "bad" way, it shouldn't matter: the sources are always identical.

File: src/engine/SCons/SConf.py [hunk 2]
```diff
@@ -592,8 +593,30 @@ def TryBuild(self, builder, text = None, extension = ""):
...
+
+            f = "_".join([f, textSig, textSigCounter])
+            textFile = self.confdir.File(f + extension)
+            textFileNode = self.env.SConfSourceBuilder(target=textFile,
+                                                       source=sourcetext)
+            nodesToBeBuilt.extend(textFileNode)
+
+            source = textFile
+            target = textFile.File(f + "SConfActionsContentDummyTarget")
```
Comment (acmorrow) [inline src/engine/SCons/SConf.py:611]: This is unfortunate, but without it, we don't have a real `Node` to pass as the `target` argument for the `get_contents` call below, and if there are any emitters in play that want to actually examine the targets, they will fall over. I'm open to suggestions on the name or other approaches to ensure that there is a real `Node` available for use as the `target`.

File: src/engine/SCons/SConf.py [hunk 3]
```diff
@@ -592,8 +593,30 @@ def TryBuild(self, builder, text = None, extension = ""):
...
+        else:
+            source = None
```
Comment (acmorrow) [inline src/engine/SCons/SConf.py:613]: I haven't explored whether this branch ever happens, or how to disambiguate if it does.

File: src/engine/SCons/SConf.py [hunk 4]
```diff
@@ -592,8 +593,30 @@ def TryBuild(self, builder, text = None, extension = ""):
...
+            target = None
+
+        action = builder.builder.action.get_contents(target=target, source=[source], env=self.env)
```
Comment (acmorrow) [inline src/engine/SCons/SConf.py:616]: Originally, what I wanted here was to synthesize this action string *before* making the source file, and appending that to the source file name. Then everything would just flow nicely. But I can't do that, because I need the name of the source file to feed in.

**Comment by acmorrow** [comment]
@bdbaddog - Any thoughts on this proposed change? It solves a real problem for my users. We have optional modules that can be enabled for the build, and they introduce new configure tests. When those modules are subsequently disabled or opted-out, the configure tests can get confused. This leads to people putting `--config=force` in their `SCONSFLAGS`, which leads to a poor user experience because the configure tests can never be cached.

**Comment by acmorrow** [comment]
@bdbaddog - I've rebased this onto master and force updated the branch to get a new test run. I took a look at some of the old test failures, and it seems that most of them are caused by the fact that the name of configuration test files are no longer predictable. Many tests are looking for `conftest_0` and failing because that file now has a hash that can't be known ahead of time attached. Any suggestions on how to approach that?

**Comment by bdbaddog** [comment]
@acmorrow - the file names should still be deterministic right? So I'd say figure out what the file name should be going forward and replace conftest_0 with that?

**Comment by bdbaddog** [comment]
@acmorrow - hop on Discord if you want to discuss..

File: src/engine/SCons/SConf.py (hunk 1 above)
Comment (bdbaddog) [reply src/engine/SCons/SConf.py:602]: Perhaps the answer then not to have different source filenames, but different target file names. This would avoid `"scons: warning: Two different environments were specified for target"`

Comment (acmorrow) [reply src/engine/SCons/SConf.py:602]: @bdbaddog - We collide on the `target` nodes of `SConfSourceBuilder`, so that is what this is actually doing: producing different target names.

**Comment by bdbaddog** [comment]
Some notes on this PR:

1. The PR does improve the current state of configure logic when tests may be run or not run or run in different order on different runs. There were minimally cases where the results should have been cached from a previous run and were rerun. And maximally cases where different runs might reuse incorrect object files to link, yielding false negatives or positives.  In those cases a rerun with ```--config=force``` and/or cleaning the configure temp file directory would repair this issue. An example false failure can occur when there's a .o generated to check for compile which has no main and thus when link is attempted will fail and the check will (possibly) fail incorrectly
2. Even with the logic in this PR there are a couple corner cases where a false positive or negative can occur (though once again cleaning the configure output files/directory and/or ```--config=force``` would resolve them).  Generally these corner cases fall into multi step builders.  Such as Program which is called via TryLink (among other checkers).  Where the source file is named according to it's content hash and the link action hash.  But there's no uniqueness provided against the expanded CCCOM. So two calls with different environments where the CCCOM is different, but not the check being called or the LINKCOM could yield a file naming conflict and a false positive or negative on the configure check.
3. It may be possible to get rid of the file naming counter if the ```SConfSourceBuilder()``` can be called either from a single Environment() (not from the env passed to Configure) or from DefaultEnvironment().

That said the improvements in this PR are worthy of merging. 
Per discussion with @acmorrow I'll be pulling the logic from this PR and adding some tests and submitting a new PR.

Additionally I'll try to quantify with some tests item 2 above where a corner case might fail this logic and create an issue for that.

**Comment by bdbaddog** [comment]
Work continued in PR #3543
